### ISSUE TITLE
More redundant_closure_for_method_calls

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#11017]
### Summary

Here I suggest one more example for the redundant_closure_for_method_calls.

### Lint Name

redundant_closure_for_method_calls

### Reproducer

I tried this code:

```rust
#![feature(slice_group_by)]
#![warn(clippy::pedantic)]
fn main() {
    let data = [true, false, true, false, true, true, false, false, true];
    let _gs1 = data.group_by(|a, b| a == b);
}
```

I expected to see this happen:
I expected a link firing like:
note: `#[warn(clippy::redundant_closure_for_method_calls)]` implied by `#[warn(clippy::pedantic)]`
Instead, this happened:
Nothing spotted by clippy. I expected Clippy to suggest me to use:

data.group_by(bool::eq);

### Version

```text
rustc 1.72.0-nightly (46514218f 2023-06-20)
binary: rustc
commit-hash: 46514218f6f31ad3a1510ecc32af47e9e486c27d
commit-date: 2023-06-20
host: x86_64-pc-windows-gnu
release: 1.72.0-nightly
LLVM version: 16.0.5
```

**Comment by zaneduffield** [comment]
The lint doesn't even work when you make it more obvious:
```rust
data.group_by(|a, b| bool::eq(a, b));
```

I can see that the check that's blocking it is right here
https://github.com/rust-lang/rust-clippy/blob/1d334696587ac22b3a9e651e7ac684ac9e0697b2/clippy_lints/src/eta_reduction.rs#L115C42-L115C42

which was added by #7661 to fix false positives in other cases. The comment above that line does not inspire confidence 🙃

I don't know enough about this to figure out what condition would exclude all the existing false positives while including this new case, but maybe this information will point someone else in the right direction 🤷

**Comment by Jarcho** [comment]
The issue here is how the compiler implements the distinction between late bound and early bound lifetimes on function definitions. I'll get back to fixing this once #8685 is merged.
